communications/mqtt/ESP32_MQTT/ESP32_MQTT.ino

```diff
@@ -0,0 +1,96 @@
+#include <WiFi.h>
+#include <PubSubClient.h>
+
+
+// Update these with values suitable for your network.
+const char* ssid = "WIFI_SSID";
+const char* password = "WIFI_Passw";
+const char* mqtt_server = "Server_IP";
+#define mqtt_port 1883
+#define MQTT_USER "User"
+#define MQTT_PASSWORD "Password"
+#define MQTT_SERIAL_PUBLISH_CH "MQTT/PUB/topic"
+#define MQTT_SERIAL_RECEIVER_CH "MQTT/SUB/topic"
+
+WiFiClient wifiClient;
+
+PubSubClient client(wifiClient);
+
+void setup_wifi() {
+    delay(10);
+    // We start by connecting to a WiFi network
+    Serial.println();
+    Serial.print("Connecting to ");
+    Serial.println(ssid);
+    WiFi.begin(ssid, password);
+    while (WiFi.status() != WL_CONNECTED) {
+      delay(500);
+      Serial.print(".");
+    }
+    randomSeed(micros());
+    Serial.println("");
+    Serial.println("WiFi connected");
+    Serial.println("IP address: ");
+    Serial.println(WiFi.localIP());
+}
+
+void reconnect() {
+  // Loop until we're reconnected
+  while (!client.connected()) {
+    Serial.print("Attempting MQTT connection...");
+    // Create a random client ID
+    String clientId = "ESP32Client-";
+    clientId += String(random(0xffff), HEX);
+    // Attempt to connect
+    if (client.connect(clientId.c_str(),MQTT_USER,MQTT_PASSWORD)) {
+      Serial.println("connected");
+      //Once connected, publish an announcement...
+      client.publish(MQTT_SERIAL_PUBLISH_CH, "hello world");
+      // ... and resubscribe
+      client.subscribe(MQTT_SERIAL_RECEIVER_CH);
+    } else {
+      Serial.print("failed, rc=");
+      Serial.print(client.state());
+      Serial.println(" try again in 5 seconds");
+      // Wait 5 seconds before retrying
+      delay(5000);
+    }
+  }
+}
+
+//We create a callback function to print the SUB topic information from the broker
+void callback(char* topic, byte *payload, unsigned int length) {
+    Serial.println("-------new message from broker-----");
+    Serial.print("channel:");
+    Serial.println(topic);
+    Serial.print("data:");  
+    Serial.write(payload, length);
+    Serial.println();
+}
+
+void setup() {
+  Serial.begin(115200);
+  Serial.setTimeout(500);// Set time out for 
+  setup_wifi();
+  client.setServer(mqtt_server, mqtt_port);
+  client.setCallback(callback);
+  reconnect();
+}
+
+//After connecting, we can publish messages to the the PUB topic by the Serial Monitor
+void publishSerialData(char *serialData){
+  if (!client.connected()) {
+    reconnect();
+  }
+  client.publish(MQTT_SERIAL_PUBLISH_CH, serialData);
+}
+
+void loop() {
+   client.loop();
+   if (Serial.available() > 0) {
+     char mun[501];
+     memset(mun,0, 501);
+     Serial.readBytesUntil( '\n',mun,500);
```

Comment:
not very clar what is mun [501], or Serial.readBytesUntil... better not to use them

Reply:
I want to see how to send a json message use this example to integrate it 

```cpp
String payload = "{\"counter\":"+String(counter++)+"}";
client.publish("counter", payload.c_str());
```

but I would like to use a library that halps to prepare the messages